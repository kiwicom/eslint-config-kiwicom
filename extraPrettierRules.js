// @flow

module.exports = {
  ...require('eslint-config-prettier').rules,
  ...require('eslint-config-prettier/flowtype').rules,
  ...require('eslint-config-prettier/react').rules,
};
