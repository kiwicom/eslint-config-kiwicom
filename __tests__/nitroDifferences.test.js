// @flow

import { diff } from 'deep-diff';

import deprecatedRules from './deprecatedRules';
import extraPrettierRules from '../extraPrettierRules';

const CLIEngine = require('eslint').CLIEngine;

expect.addSnapshotSerializer({
  print(serializerValue) {
    return serializerValue;
  },
  test() {
    return true;
  },
});

const explainedDifferencies = new Set([
  // Rules 'no-unused-expressions' and 'babel/no-unused-expressions' are broken and
  // replaced by 'flow/no-unused-expressions' (it supports not only optional chaining
  // but also exhaustive type checking).
  'no-unused-expressions',

  // This rule is outdated and doesn't sort Flow types correctly (as we like).
  // See: https://gitlab.skypicker.com/incubator/universe/commit/b1d0be4c7b4bfb6fa3557694f7cfcf0bfde05dfb
  // See: https://github.com/airbnb/javascript/blob/4539dbcf17b960ab650665bc95421dc8d0d3276e/packages/eslint-config-airbnb/rules/react.js#L241
  'react/sort-comp',
]);

/**
 * N - indicates a newly added property/element
 * D - indicates a property/element was deleted
 * E - indicates a property/element was edited
 * A - indicates a change occurred within an array
 *
 * See: https://github.com/flitbit/diff#differences
 */
test('differences with @kiwicom/eslint-config-nitro config', () => {
  const consoleSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});

  expect(
    formatChanges(
      'This changelog describes what should Universe change in order to be identical with Nitro config:',
      diff(getRules('../strict'), getRules('@kiwicom/eslint-config-nitro')),
    ),
  ).toMatchSnapshot();

  // This reversed changed is not just naively changed direction of changes.
  // For example, some rules should be added in one direction but NOT removed
  // in the other direction (deletions are not reflected at all).
  expect(
    formatChanges(
      'This changelog describes what should Nitro change in order to be identical with Universe config:',
      diff(getRules('@kiwicom/eslint-config-nitro'), getRules('../strict')),
    ),
  ).toMatchSnapshot();

  expect(consoleSpy.mock.calls).toEqual([
    ["! Nitro Eslint config is deprecated. Use '@kiwicom/eslint-config/nitro' instead."],
  ]);
  consoleSpy.mockRestore();
});

function formatChanges(description, changes) {
  let differences = `\n${description}\n\n`;
  changes.forEach(change => {
    const { kind } = change;
    let context = '';
    if (kind === 'E') {
      context += `(${JSON.stringify(change.lhs)} -> ${JSON.stringify(change.rhs)})`;
    } else if (kind === 'A') {
      context += `(${JSON.stringify(change.item)})`;
    } else if (kind === 'N') {
      context += `(${JSON.stringify(change.rhs)})`;
    }
    if (kind !== 'D') {
      // We do not add DELETED rules into final diff snapshot. This means that we
      // have some additional rules Nitro doesn't contain and there is nothing
      // to do from our side (Nitro should add them?).
      differences += `${kind} - ${change.path} ${context}\n`;
    }
  });
  return differences;
}

function normalizeLevel(level) {
  if (level === 2) {
    return 'error';
  } else if (level === 1) {
    return 'warn';
  } else if (level === 0) {
    return 'off';
  }
  return level;
}

function normalizeArrayConfig(config) {
  // "off" rule is disabled no matter what is the config (so we remove the config)
  const arrayConfig = config[0] === 'off' ? 'off' : config;
  return arrayConfig.length === 1 ? arrayConfig[0] : arrayConfig;
}

function normalize(config) {
  return Object.entries(config).reduce((acc, [ruleName, value]) => {
    let normalizedConfig;
    if (Array.isArray(value)) {
      const [level, ...config] = value;
      normalizedConfig = normalizeArrayConfig([normalizeLevel(level), ...config]);
    } else {
      normalizedConfig = normalizeLevel(value);
    }
    if (
      !deprecatedRules.has(ruleName) &&
      !explainedDifferencies.has(ruleName) &&
      !(ruleName in extraPrettierRules)
    ) {
      // Add only rules which are not deprecated or excluded from the diff.
      // We also skip rules conflicting with prettier (they should always be disabled).
      acc[ruleName] = normalizedConfig;
    }
    return acc;
  }, {});
}

function getRules(dependency) {
  // $FlowAllowDynamicImport
  const dependencyFile = require(dependency);
  const dependencyConfig = new CLIEngine({
    useEslintrc: false,
    rules: dependencyFile.rules,
    baseConfig: {
      extends: dependencyFile.extends,
    },
  }).getConfigForFile(__filename);
  return normalize(dependencyConfig.rules);
}
