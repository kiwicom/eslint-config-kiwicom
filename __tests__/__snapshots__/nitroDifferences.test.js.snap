// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`differences with @kiwicom/eslint-config-nitro config 1`] = `
E - getter-return ("error" vs. ["error",{"allowImplicit":true}])
E - no-async-promise-executor ("warn" vs. "off")
E - no-cond-assign ("error" vs. ["error","always"])
E - no-console ("error" vs. "warn")
E - no-constant-condition (["warn",{"checkLoops":false}] vs. "warn")
E - no-empty (["warn",{"allowEmptyCatch":true}] vs. "error")
E - no-inner-declarations ("warn" vs. "error")
E - no-misleading-character-class ("error" vs. "off")
E - no-prototype-builtins ("off" vs. "error")
E - no-template-curly-in-string ("off" vs. "error")
E - require-atomic-updates ("error" vs. "off")
E - valid-typeof ("off" vs. ["error",{"requireStringLiterals":true}])
E - accessor-pairs (["warn",{"setWithoutGet":true,"getWithoutSet":false}] vs. "off")
E - array-callback-return ("off" vs. ["error",{"allowImplicit":true}])
E - block-scoped-var ("off" vs. "error")
E - class-methods-use-this ("off" vs. ["error",{"exceptMethods":["render","getInitialState","getDefaultProps","getChildContext","componentWillMount","UNSAFE_componentWillMount","componentDidMount","componentWillReceiveProps","UNSAFE_componentWillReceiveProps","shouldComponentUpdate","componentWillUpdate","UNSAFE_componentWillUpdate","componentDidUpdate","componentWillUnmount","componentDidCatch","getSnapshotBeforeUpdate"]}])
E - consistent-return (["error",{"treatUndefinedAsUnspecified":false}] vs. "error")
E - default-case ("off" vs. ["error",{"commentPattern":"^no default$"}])
E - dot-notation ("warn" vs. ["error",{"allowKeywords":true,"allowPattern":""}])
A - eqeqeq ({"kind":"N","rhs":{"null":"ignore"}})
E - eqeqeq,1 ("smart" vs. "always")
E - eqeqeq,0 ("warn" vs. "error")
E - guard-for-in ("off" vs. "error")
E - no-await-in-loop ("warn" vs. "error")
E - no-case-declarations ("warn" vs. "error")
E - no-div-regex ("warn" vs. "off")
E - no-else-return ("error" vs. ["error",{"allowElseIf":false}])
E - no-empty-function ("off" vs. ["error",{"allow":["arrowFunctions","functions","methods"]}])
E - no-empty-pattern ("warn" vs. "error")
E - no-extend-native ("warn" vs. "error")
E - no-extra-bind ("warn" vs. "error")
E - no-extra-label ("warn" vs. "error")
E - no-fallthrough ("warn" vs. "error")
A - no-global-assign,1,exceptions ({"kind":"D","lhs":"Set"})
A - no-global-assign,1,exceptions ({"kind":"D","lhs":"Map"})
E - no-iterator ("warn" vs. "error")
E - no-labels,1,allowLoop (true vs. false)
E - no-labels,1,allowSwitch (true vs. false)
E - no-lone-blocks ("warn" vs. "error")
E - no-loop-func ("off" vs. "error")
E - no-new-wrappers ("warn" vs. "error")
E - no-new ("warn" vs. "error")
E - no-octal ("warn" vs. "error")
E - no-octal-escape ("warn" vs. "error")
E - no-param-reassign ("error" vs. ["error",{"props":true,"ignorePropertyModificationsFor":["acc","accumulator","e","ctx","req","request","res","response","$scope"]}])
E - no-redeclare (["warn",{"builtinGlobals":true}] vs. "error")
E - no-restricted-properties ("off" vs. ["error",{"object":"arguments","property":"callee","message":"arguments.callee is deprecated"},{"object":"global","property":"isFinite","message":"Please use Number.isFinite instead"},{"object":"self","property":"isFinite","message":"Please use Number.isFinite instead"},{"object":"window","property":"isFinite","message":"Please use Number.isFinite instead"},{"object":"global","property":"isNaN","message":"Please use Number.isNaN instead"},{"object":"self","property":"isNaN","message":"Please use Number.isNaN instead"},{"object":"window","property":"isNaN","message":"Please use Number.isNaN instead"},{"property":"__defineGetter__","message":"Please use Object.defineProperty instead."},{"property":"__defineSetter__","message":"Please use Object.defineProperty instead."},{"object":"Math","property":"pow","message":"Use the exponentiation operator (**) instead."}])
E - no-return-assign ("warn" vs. ["error","always"])
E - no-return-await ("warn" vs. "error")
E - no-self-assign ("warn" vs. ["error",{"props":false}])
E - no-self-compare ("warn" vs. "error")
E - no-sequences ("warn" vs. "error")
E - no-throw-literal ("warn" vs. "error")
E - no-unused-expressions ("off" vs. ["error",{"allowShortCircuit":false,"allowTernary":false,"allowTaggedTemplates":false}])
E - no-unused-labels ("warn" vs. "error")
E - no-useless-call ("warn" vs. "off")
E - no-useless-concat ("warn" vs. "error")
E - no-useless-escape ("warn" vs. "error")
E - no-useless-return ("warn" vs. "error")
E - no-void ("warn" vs. "error")
E - no-with ("warn" vs. "error")
E - prefer-promise-reject-errors ("warn" vs. ["error",{"allowEmptyReject":true}])
E - require-await ("error" vs. "off")
E - vars-on-top ("off" vs. "error")
E - yoda ("warn" vs. "error")
E - strict ("warn" vs. ["error","never"])
E - no-restricted-globals ("off" vs. ["error","isFinite","isNaN","addEventListener","blur","close","closed","confirm","defaultStatus","event","external","defaultstatus","find","focus","frameElement","frames","history","innerHeight","innerWidth","length","location","locationbar","menubar","moveBy","moveTo","name","onblur","onerror","onfocus","onload","onresize","onunload","open","opener","opera","outerHeight","outerWidth","pageXOffset","pageYOffset","parent","print","removeEventListener","resizeBy","resizeTo","screen","screenLeft","screenTop","screenX","screenY","scroll","scrollbars","scrollBy","scrollTo","scrollX","scrollY","self","status","statusbar","stop","toolbar","top"])
E - no-shadow-restricted-names ("warn" vs. "error")
E - no-shadow ("off" vs. "error")
E - no-undef-init ("off" vs. "error")
N - no-unused-vars,1,vars ("all")
N - no-unused-vars,1,ignoreRestSiblings (true)
E - no-use-before-define ("off" vs. ["error",{"functions":true,"classes":true,"variables":true}])
E - global-require ("off" vs. "error")
E - handle-callback-err ("warn" vs. "off")
E - no-buffer-constructor ("warn" vs. "error")
E - no-new-require ("warn" vs. "error")
E - no-path-concat ("warn" vs. "error")
E - func-names ("off" vs. "warn")
E - lines-between-class-members,2,exceptAfterSingleLine (true vs. false)
E - lines-between-class-members,0 ("warn" vs. "error")
N - new-cap,1,newIsCapExceptions ([])
N - new-cap,1,capIsNewExceptions (["Immutable.Map","Immutable.Set","Immutable.List"])
E - new-cap,0 ("warn" vs. "error")
E - no-array-constructor ("warn" vs. "error")
E - no-bitwise ("warn" vs. "error")
E - no-continue ("off" vs. "error")
E - no-multi-assign ("off" vs. ["error"])
E - no-nested-ternary ("off" vs. "error")
E - no-new-object ("warn" vs. "error")
E - no-plusplus ("off" vs. "error")
E - no-restricted-syntax ("off" vs. ["error",{"selector":"ForInStatement","message":"for..in loops iterate over the entire prototype chain, which is virtually never what you want. Use Object.{keys,values,entries}, and iterate over the resulting array."},{"selector":"ForOfStatement","message":"iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations."},{"selector":"LabeledStatement","message":"Labels are a form of GOTO; using them makes code confusing and hard to maintain and understand."},{"selector":"WithStatement","message":"\`with\` is disallowed in strict mode because it makes code impossible to predict and optimize."}])
E - no-underscore-dangle ("off" vs. ["error",{"allow":[],"allowAfterThis":false,"allowAfterSuper":false,"enforceInMethodNames":false}])
E - no-unneeded-ternary ("warn" vs. ["error",{"defaultAssignment":false}])
E - one-var,1 ({"initialized":"never"} vs. "never")
E - one-var,0 ("warn" vs. "error")
E - operator-assignment ("warn" vs. ["error","always"])
E - spaced-comment,2,block,markers,1 ("::" vs. "!")
E - spaced-comment,2,block,markers,0 (":" vs. "=")
N - spaced-comment,2,block,exceptions (["-","+"])
N - spaced-comment,2,line ({"exceptions":["-","+"],"markers":["=","!"]})
E - no-class-assign ("warn" vs. "error")
E - no-new-symbol ("warn" vs. "error")
E - no-useless-computed-key ("warn" vs. "error")
E - no-useless-constructor ("warn" vs. "error")
E - no-var ("warn" vs. "error")
E - no-useless-rename ("warn" vs. ["error",{"ignoreDestructuring":false,"ignoreImport":false,"ignoreExport":false}])
E - object-shorthand ("off" vs. ["error","always",{"ignoreConstructors":false,"avoidQuotes":true}])
E - prefer-const,1,destructuring ("all" vs. "any")
E - prefer-const,1,ignoreReadBeforeAssign (false vs. true)
E - prefer-destructuring ("off" vs. ["error",{"VariableDeclarator":{"array":false,"object":true},"AssignmentExpression":{"array":true,"object":true}},{"enforceForRenamedProperties":false}])
E - prefer-numeric-literals ("off" vs. "error")
E - prefer-rest-params ("warn" vs. "error")
E - prefer-spread ("warn" vs. "error")
E - prefer-template ("off" vs. "error")
E - require-yield ("off" vs. "error")
E - symbol-description ("warn" vs. "error")
A - flowtype/require-valid-file-annotation ({"kind":"N","rhs":{"annotationStyle":"line"}})
E - react/button-has-type ("off" vs. ["error",{"button":true,"submit":true,"reset":false}])
E - react/destructuring-assignment ("off" vs. ["error","always"])
E - react/forbid-dom-props (["error",{"forbid":["class","for"]}] vs. "off")
E - react/forbid-foreign-prop-types ("off" vs. ["warn",{"allowInPropTypes":true}])
E - react/forbid-prop-types ("off" vs. ["error",{"forbid":["any","array","object"],"checkContextTypes":true,"checkChildContextTypes":true}])
E - react/jsx-boolean-value ("off" vs. ["error","never",{"always":[]}])
E - react/jsx-curly-brace-presence ("off" vs. ["error",{"props":"never","children":"never"}])
E - react/jsx-filename-extension ("off" vs. ["error",{"extensions":[".jsx"]}])
E - react/jsx-key ("error" vs. "off")
E - react/jsx-no-bind ("error" vs. ["error",{"ignoreRefs":true,"allowArrowFunctions":true,"allowFunctions":false,"allowBind":false,"ignoreDOMComponents":true}])
E - react/jsx-no-comment-textnodes ("warn" vs. "error")
E - react/jsx-no-duplicate-props ("error" vs. ["error",{"ignoreCase":true}])
E - react/jsx-no-target-blank ("off" vs. ["error",{"enforceDynamicLinks":"always"}])
E - react/jsx-pascal-case ("off" vs. ["error",{"allowAllCaps":true,"ignore":[]}])
E - react/jsx-uses-react ("warn" vs. ["error"])
E - react/jsx-uses-vars ("warn" vs. "error")
E - react/no-array-index-key ("off" vs. "error")
E - react/no-children-prop ("off" vs. "error")
E - react/no-danger ("off" vs. "warn")
E - react/no-danger-with-children ("off" vs. "error")
E - react/no-deprecated ("error" vs. ["error"])
E - react/no-direct-mutation-state ("error" vs. "off")
E - react/no-find-dom-node ("off" vs. "error")
E - react/no-is-mounted ("warn" vs. "error")
E - react/no-multi-comp ("off" vs. ["error",{"ignoreStateless":true}])
E - react/no-render-return-value ("off" vs. "error")
E - react/no-string-refs ("off" vs. "error")
E - react/no-this-in-sfc ("off" vs. "error")
E - react/no-typos ("off" vs. "error")
E - react/no-unescaped-entities ("off" vs. "error")
E - react/no-unknown-property ("warn" vs. "error")
E - react/no-unused-prop-types ("off" vs. ["error",{"customValidators":[],"skipShapeProps":true}])
E - react/no-unused-state ("off" vs. "error")
E - react/no-will-update-set-state ("off" vs. "error")
E - react/prefer-es6-class ("off" vs. ["error","always"])
E - react/prefer-stateless-function ("off" vs. ["error",{"ignorePureComponents":true}])
E - react/prop-types ("off" vs. ["error",{"ignore":[],"customValidators":[],"skipUndeclared":false}])
E - react/react-in-jsx-scope ("warn" vs. "error")
A - react/sort-comp,1,order ({"kind":"N","rhs":"rendering"})
A - react/sort-comp,1,order ({"kind":"N","rhs":"everything-else"})
A - react/sort-comp,1,order ({"kind":"N","rhs":"instance-methods"})
A - react/sort-comp,1,order ({"kind":"N","rhs":"/^(get|set)(?!(InitialState$|DefaultProps$|ChildContext$)).+$/"})
E - react/sort-comp,1,order,5 ("rendering" vs. "setters")
E - react/sort-comp,1,order,4 ("everything-else" vs. "getters")
E - react/sort-comp,1,order,3 ("lifecycle" vs. "/^on.+$/")
E - react/sort-comp,1,order,2 ("instance-variables" vs. "lifecycle")
E - react/sort-comp,1,order,1 ("static-methods" vs. "instance-variables")
E - react/sort-comp,1,order,0 ("type-annotations" vs. "static-methods")
A - react/sort-comp,1,groups,lifecycle ({"kind":"D","lhs":"componentWillUnmount"})
A - react/sort-comp,1,groups,lifecycle ({"kind":"D","lhs":"componentDidCatch"})
A - react/sort-comp,1,groups,lifecycle ({"kind":"D","lhs":"componentDidUpdate"})
A - react/sort-comp,1,groups,lifecycle ({"kind":"D","lhs":"getSnapshotBeforeUpdate"})
A - react/sort-comp,1,groups,lifecycle ({"kind":"D","lhs":"UNSAFE_componentWillUpdate"})
A - react/sort-comp,1,groups,lifecycle ({"kind":"D","lhs":"componentWillUpdate"})
A - react/sort-comp,1,groups,lifecycle ({"kind":"D","lhs":"shouldComponentUpdate"})
E - react/sort-comp,1,groups,lifecycle,18 ("UNSAFE_componentWillReceiveProps" vs. "componentWillUnmount")
E - react/sort-comp,1,groups,lifecycle,17 ("componentWillReceiveProps" vs. "componentDidUpdate")
E - react/sort-comp,1,groups,lifecycle,16 ("componentDidMount" vs. "componentWillUpdate")
E - react/sort-comp,1,groups,lifecycle,15 ("UNSAFE_componentWillMount" vs. "shouldComponentUpdate")
E - react/sort-comp,1,groups,lifecycle,14 ("componentWillMount" vs. "componentWillReceiveProps")
E - react/sort-comp,1,groups,lifecycle,13 ("getDerivedStateFromError" vs. "componentDidMount")
E - react/sort-comp,1,groups,lifecycle,12 ("getDerivedStateFromProps" vs. "componentWillMount")
E - react/sort-comp,1,groups,lifecycle,10 ("getInitialState" vs. "state")
E - react/sort-comp,1,groups,lifecycle,9 ("state" vs. "getInitialState")
E - react/style-prop-object ("off" vs. "error")
E - react/void-dom-elements-no-children ("off" vs. "error")
E - jsx-a11y/alt-text ("error" vs. ["error",{"elements":["img","object","area","input[type=\\"image\\"]"],"img":[],"object":[],"area":[],"input[type=\\"image\\"]":[]}])
E - jsx-a11y/anchor-has-content ("off" vs. ["error",{"components":[]}])
E - jsx-a11y/anchor-is-valid ("error" vs. "off")
E - jsx-a11y/aria-activedescendant-has-tabindex ("off" vs. "error")
E - jsx-a11y/aria-role ("error" vs. ["error",{"ignoreNonDom":false,"ignoreNonDOM":false}])
E - jsx-a11y/click-events-have-key-events ("off" vs. "error")
E - jsx-a11y/heading-has-content ("off" vs. ["error",{"components":[""]}])
E - jsx-a11y/img-redundant-alt ("off" vs. "error")
E - jsx-a11y/label-has-associated-control ("off" vs. ["error",{"labelComponents":[],"labelAttributes":[],"controlComponents":[],"assert":"both","depth":25}])
E - jsx-a11y/label-has-for ("off" vs. ["error",{"components":[],"required":{"every":["nesting","id"]},"allowChildren":false}])
E - jsx-a11y/media-has-caption ("off" vs. ["error",{"audio":[],"video":[],"track":[]}])
E - jsx-a11y/mouse-events-have-key-events ("off" vs. "error")
E - jsx-a11y/no-interactive-element-to-noninteractive-role ("off" vs. ["error",{"tr":["none","presentation"]}])
E - jsx-a11y/no-noninteractive-element-interactions ("off" vs. ["error",{"handlers":["onClick","onMouseDown","onMouseUp","onKeyPress","onKeyDown","onKeyUp"]}])
E - jsx-a11y/no-noninteractive-element-to-interactive-role ("off" vs. ["error",{"ul":["listbox","menu","menubar","radiogroup","tablist","tree","treegrid"],"ol":["listbox","menu","menubar","radiogroup","tablist","tree","treegrid"],"li":["menuitem","option","row","tab","treeitem"],"table":["grid"],"td":["gridcell"]}])
E - jsx-a11y/no-noninteractive-tabindex ("off" vs. ["error",{"tags":[],"roles":["tabpanel"]}])
E - jsx-a11y/no-static-element-interactions ("off" vs. ["error",{"handlers":["onClick","onMouseDown","onMouseUp","onKeyPress","onKeyDown","onKeyUp"]}])
E - relay/compat-uses-vars ("off" vs. "warn")
E - relay/generated-flow-types ("error" vs. "warn")
E - relay/no-future-added-value ("error" vs. "warn")
E - relay/unused-fields ("error" vs. "warn")
E - import/extensions ("off" vs. ["error","ignorePackages",{"js":"never","mjs":"never","jsx":"never"}])
E - import/named ("off" vs. "error")
E - import/no-anonymous-default-export (["error",{"allowArray":true,"allowArrowFunction":false,"allowAnonymousClass":false,"allowAnonymousFunction":false,"allowCallExpression":true,"allowLiteral":true,"allowObject":true}] vs. "off")
E - import/no-cycle ("off" vs. ["error",{"maxDepth":null}])
E - import/no-duplicates ("warn" vs. "error")
E - import/no-dynamic-require ("off" vs. "error")
A - import/no-extraneous-dependencies,1,devDependencies ({"kind":"N","rhs":"**/__tests__/*"})
E - import/no-extraneous-dependencies,1,devDependencies,1 ("**/*.spec.js" vs. "stories/**")
E - import/no-extraneous-dependencies,1,devDependencies,0 ("**/*.test.js" vs. "etc/**")
E - import/no-named-as-default-member ("off" vs. "error")
N - import/no-unresolved,1,commonjs (true)
N - import/no-unresolved,1,caseSensitive (true)
A - import/order,1,groups ({"kind":"D","lhs":"index"})
A - import/order,1,groups,1 ({"kind":"N","rhs":"index"})
A - import/order,1,groups,1 ({"kind":"N","rhs":"sibling"})
E - import/order,1,groups,1,1 ("sibling" vs. "parent")
E - import/order,1,groups,1,0 ("parent" vs. "internal")
E - import/prefer-default-export ("off" vs. "error")
E - flowtype/boolean-style ("off" vs. ["error","boolean"])
E - flowtype/generic-spacing ("off" vs. ["error","never"])
E - flowtype/space-before-generic-bracket ("off" vs. ["error","never"])
E - flowtype/space-before-type-colon ("off" vs. ["error","never"])
E - flowtype/union-intersection-spacing ("off" vs. ["error","always"])
E - prettier/prettier,1,printWidth (80 vs. 100)
N - lines-around-directive (["error",{"before":"always","after":"always"}])
N - react/jsx-sort-prop-types ("off")
N - fp/no-nil ("off")
N - fp/no-class ("off")
N - fp/no-this ("off")
N - fp/no-mutation ("off")
N - fp/no-unused-expression ("off")

`;
